File: C#/Queues/CircularQueue/CircularQueue.cs
/*******************************************************
 *  CircularQueue.cs
 *  Created by Stephen Hall on 11/20/17.
 *  Copyright (c) 2017 Stephen Hall. All rights reserved.
 *  Circular Queue implementation in C#
 ********************************************************/

namespace DataStructures.Queues.CircularQueue
{
    /// <summary>
    /// Circular Queue Class
    /// </summary>
    /// <typeparam name="T">Generic type</typeparam>
    public class CircularQueue<T>
    {
        /// <summary>
        /// Private Members
        /// </summary>
        private T[] _array;
        private int _count;
        private int _size;
        private int _zeroIndex;
       
        /// <summary>
        /// Default Constructor
        /// </summary>
        public CircularQueue()
        {
            _array = new T[_size = 10];
            _count = _zeroIndex = 0;
        }

        /// <summary>
        /// Circular Queue Constructor
        /// </summary>
        /// <param name="size">Size to initialize the queue to</param>
        public CircularQueue(int size)
        {
            _array = new T[_size = size];
            _count = _zeroIndex = 0;
        }

        /// <summary>
        /// Pushes given data onto the queue if space is available
        /// </summary>
        /// <param name="data">Data to be added to the queue</param>
        /// <returns>Node added to the queue</returns>
        public T Enqueue(T data)
        {
            if (!IsFull())
            {
                _array[(_zeroIndex + _count) % _size] = data;
                _count++;
                return Top();
            }
            return default(T);
        }

        /// <summary>
        /// Removes item from the queue
        /// </summary>
        /// <returns>item removed from the queue</returns>
        public T Dequeue()
        {
            if (IsEmpty())
                return default(T);

            T tmp = _array[_zeroIndex];
            _array[_zeroIndex] = default(T);
            _count--;
            _zeroIndex = (_zeroIndex + 1) % _size;
            return tmp;
        }

        /// <summary>
        /// Gets the top item of the queue
        /// </summary>
        /// <returns>item on top of the queue</returns>
        public T Top() => IsEmpty() ? default(T) : _array[_zeroIndex];

        /// <summary>
        /// Returns a value indicating if the queue is empty
        /// </summary>
        /// <returns>true if empty, false if not</returns>
        public bool IsEmpty() => _count == 0; 

        /// <summary>
        ///  * Returns a value indicating if the queue is full
        /// </summary>
        /// <returns>True if full, false if not</returns>
        public bool IsFull() => _count == _size;
    }
}

File: Kotlin/Queues/CircularQueue/CircularQueue.kt
/*******************************************************
 * CircularQueue.kt
 * Created by Stephen Hall on 11/08/18.
 * Copyright (c) 2018 Stephen Hall. All rights reserved.
 * Circular Queue implementation in Java
 ******************************************************/


/**
 * Circular Queue Class
 * @param <T> Generic type
</T> */
class CircularQueue<T>
/**
 * Circular Queue Constructor
 * @param size Size to initialize the queue to
 */
(size: Int) {
    /**
     * Private Members
     */
    private val array: Array<T?>
    private var count: Int = 0
    private var size: Int = 0
    private var zeroIndex: Int = 0

    /**
     * Returns a value indicating if the queue is empty
     * @return true if empty, false if not
     */
    val isEmpty: Boolean
        get() = count == 0

    /**
     * Returns a value indicating if the queue is full
     * @return True if full, false if not
     */
    val isFull: Boolean
        get() = count == size

    /**
     * Default Constructor
     */
    constructor() : this(10) {}

    init {
        this.size = size
        array = arrayOfNulls<Any>(size) as Array<T?>
        zeroIndex = 0
        count = zeroIndex
    }

    /**
     * Pushes given data onto the queue if space is available
     * @param data Data to be added to the queue
     * @return Node added to the queue
     */
    fun enqueue(data: T): T? {
        if (!isFull) {
            array[(zeroIndex + count) % size] = data
            count++
            return top()
        }
        return null
    }

    /**
     * Removes item from the queue
     * @return item removed from the queue
     */
    fun dequeue(): T? {
        if (isEmpty)
            return null

        val tmp = array[zeroIndex]
        array[zeroIndex] = null
        count--
        zeroIndex = (zeroIndex + 1) % size
        return tmp
    }

    /**
     * Gets the top item of the queue
     * @return item on top of the queue
     */
    fun top(): T? {
        return if (isEmpty) null else array[zeroIndex]
    }
}
